src/main/java/graphql/schema2/GraphQLDirective.java

```diff
@@ -0,0 +1,211 @@
+package graphql.schema2;
+
+
+import graphql.PublicApi;
+import graphql.schema.GraphQLArgument;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.UnaryOperator;
+
+import static graphql.Assert.assertNotNull;
+import static graphql.Assert.assertValidName;
+import static graphql.introspection.Introspection.DirectiveLocation;
+
+/**
+ * A directive can be used to modify the behavior of a graphql field.
+ *
+ * See http://graphql.org/learn/queries/#directives for more details on the concept.
+ */
+@SuppressWarnings("DeprecatedIsStillUsed") // because the graphql spec still has some of these deprecated fields
+@PublicApi
+public class GraphQLDirective {
+
+    private final String name;
+    private final String description;
+    private final EnumSet<DirectiveLocation> locations;
+    private final List<GraphQLArgument> arguments = new ArrayList<>();
+    private final boolean onOperation;
+    private final boolean onFragment;
+    private final boolean onField;
+
+    public GraphQLDirective(String name, String description, EnumSet<DirectiveLocation> locations,
+                            List<GraphQLArgument> arguments, boolean onOperation, boolean onFragment, boolean onField) {
+        assertValidName(name);
+        assertNotNull(arguments, "arguments can't be null");
+        this.name = name;
+        this.description = description;
+        this.locations = locations;
+        this.arguments.addAll(arguments);
+        this.onOperation = onOperation;
+        this.onFragment = onFragment;
+        this.onField = onField;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public List<GraphQLArgument> getArguments() {
+        return new ArrayList<>(arguments);
+    }
+
+    public GraphQLArgument getArgument(String name) {
+        for (GraphQLArgument argument : arguments) {
+            if (argument.getName().equals(name)) return argument;
+        }
+        return null;
+    }
+
+    public EnumSet<DirectiveLocation> validLocations() {
+        return locations;
+    }
+
+    /**
+     * @return onOperation
+     *
+     * @deprecated Use {@link #validLocations()}
+     */
+    @Deprecated
+    public boolean isOnOperation() {
+        return onOperation;
+    }
+
+    /**
+     * @return onFragment
+     *
+     * @deprecated Use {@link #validLocations()}
+     */
+    @Deprecated
+    public boolean isOnFragment() {
+        return onFragment;
+    }
+
+    /**
+     * @return onField
+     *
+     * @deprecated Use {@link #validLocations()}
+     */
+    @Deprecated
+    public boolean isOnField() {
+        return onField;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public static Builder newDirective() {
+        return new Builder();
+    }
+
+    public static class Builder {
```

Comment:
So if the Builders for GraphqlObjectType are external - why is this one (and others internal?)  Did you just not get around to it and plan to make it external?

I think it should be one way or the others but not a mix